packages/core/src/sync-realtime/service.ts

```diff
@@ -183,89 +187,95 @@ export const start = (service: Service) => {
+        if (latestLocalBlock.number + 1 < newHeadBlockNumber) {
+          // Retrieve missing blocks, but only fetch 50 at most.
+          const missingBlockRange = range(
+            latestLocalBlock.number + 1,
+            Math.min(newHeadBlockNumber, latestLocalBlock.number + 51),
+          );
+          const pendingBlocks = await Promise.all(
+            missingBlockRange.map((blockNumber) =>
+              _eth_getBlockByNumber(service, { blockNumber }),
+            ),
+          );
+
+          service.common.logger.debug({
+            service: "realtime",
+            msg: `Fetched ${missingBlockRange.length} missing '${
+              service.network.name
+            }' blocks from ${latestLocalBlock.number + 1} to ${Math.min(
+              newHeadBlockNumber,
+              latestLocalBlock.number + 51,
+            )}`,
+          });
 
-          // Check if a reorg occurred by validating the chain of block hashes.
-          if (newHeadBlock.parentHash !== latestLocalBlock.hash) {
-            await handleReorg(service, newHeadBlock);
-            queue.clear();
-            return;
+          // This is needed to ensure proper `kill()` behavior. When the service
+          // is killed, nothing should be added to the queue, or else `onIdle()`
+          // will never resolve.
+          if (service.isKilled) return;
+
+          queue.clear();
+
+          for (const pendingBlock of pendingBlocks) {
+            queue.add(pendingBlock);
           }
 
-          // New block is exactly one block ahead of the local chain.
-          // Attempt to ingest it.
-          await handleBlock(service, { newHeadBlock });
+          queue.add(newHeadBlock);
 
           return;
-        } catch (_error) {
-          if (service.isKilled) return;
+        }
+
+        // Check if a reorg occurred by validating the chain of block hashes.
+        if (newHeadBlock.parentHash !== latestLocalBlock.hash) {
+          await handleReorg(service, newHeadBlock);
+          queue.clear();
+          return;
+        }
 
-          const error = _error as Error;
+        // New block is exactly one block ahead of the local chain.
+        // Attempt to ingest it.
+        await handleBlock(service, { newHeadBlock });
 
-          service.common.logger.warn({
+        // Reset the error state after successfully completing the happy path.
+        service.consecutiveErrors = 0;
+
+        return;
+      } catch (_error) {
+        if (service.isKilled) return;
+
+        const error = _error as Error;
+
+        service.common.logger.warn({
+          service: "realtime",
+          msg: `Failed to process '${service.network.name}' block ${newHeadBlockNumber} with error: ${error}`,
+        });
+
+        await wait(250 * 2 ** service.consecutiveErrors);
+
+        // Remove all blocks from the queue. This protects against an
+        // erroneous block causing a fatal error.
+        queue.clear();
```

Comment:
Are there any scenarios where we'd be wasting a lot of valid "work" by doing this here?

Reply:
Yes, one example would be that the `wait` block is executing and meanwhile real blocks are getting added to the queue. Then once the `wait` finished, the real blocks are instantly cleared. We could add a `paused` state but only if you think it is worth the complexity